File: gdx2d-desktop/src/hevs/gdx2d/demos/physics/pinball/Renderer.java
package hevs.gdx2d.demos.physics.pinball;

import com.badlogic.gdx.physics.box2d.Body;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.utils.Array;

import hevs.gdx2d.lib.GdxGraphics;
import hevs.gdx2d.lib.interfaces.DrawableObject;
import hevs.gdx2d.lib.physics.PhysicsWorld;

public class Renderer {
	public void draw(GdxGraphics g) {
		World world = PhysicsWorld.getInstance();
		Array<Body> bodies = new Array<Body>(world.getBodyCount());
		world.getBodies(bodies);
		
		for (int i = 0 ; i < bodies.size; i++) {
			Object o = bodies.get(i).getUserData();
			
			if (o instanceof DrawableObject) {
				((DrawableObject) o).draw(g);
			}
		}
	}
}

File: gdx2d-demoDesktop/src/main/scala/ch/hevs/gdx2d/demos/physics/pinball/Renderer.scala
package ch.hevs.gdx2d.demos.physics.pinball

import java.util.LinkedList

import ch.hevs.gdx2d.lib.GdxGraphics
import ch.hevs.gdx2d.lib.interfaces.DrawableObject
import com.badlogic.gdx.physics.box2d.{Body, World}
import com.badlogic.gdx.utils.Array

class Renderer {
  def draw(g: GdxGraphics, world: World, decorations: LinkedList[TemporaryDrawable]): Unit = {
    val bodies = new Array[Body](world.getBodyCount)
    world.getBodies(bodies)

    // Draw physical objects
    for (i <- 0 until bodies.size) {
      val o = bodies.get(i).getUserData
      if (o.isInstanceOf[DrawableObject]) {
        o.asInstanceOf[DrawableObject].draw(g)
      }
    }

    // Draw decorations
    val toRemove = new LinkedList[TemporaryDrawable]()
    val it = decorations.iterator()
    while (it.hasNext) {
      val d = it.next()
      d.draw(g)
      if (d.isDone) {
        toRemove.add(d)
      }
    }

    decorations.removeAll(toRemove)
  }
}
